lib/experimental/class-wp-html-processor.php

```diff
@@ -0,0 +1,720 @@
+ *    some fairly robust support because it's not difficult to
+ *    create misleading HTML through the block editor. For
+ *    example, as can happen with a number of plaintext inputs.
+ *
+ * Performance ideas:
+ *
+ * It may be possible that we're thrashing the garbage collector by
+ * allocating new classes for each tag and attribute token as we
+ * scan through the input document. If this is the case we could
+ * try pre-allocating a certain number of token class instances and
+ * reuse those upon scanning each new tag. This would incur a fixed
+ * memory cost but in addition to lowering the pressure on the
+ * garbage collector we might end up having more efficient use of
+ * localized data within the processor cache.
+ *
+ * If we can pick up scanning where we leave off, we can stop parsing
+ * within a tag opener as soon as we meet all the query constraints.
+ * If we need to set an attribute we haven't yet seen then we would
+ * need to continue scanning to the end of that tag opener until we
+ * find the existing attribute or know it's not there. This could
+ * avoid parsing large attributes, such as one with a long data-URI.
+ *
+ * Defer creating comparable values until we need them so we don't
+ * eagerly allocate and convert string values we don't ever read.
+ *
+ * Classes are passed by reference by default, but are we missing
+ * opportunities to pass by reference when we are copying by value?
+ */
+
+
+/**
+ * Processes an input HTML document by applying a specified set
+ * of patches to that input. Tokenizes HTML but does not fully
+ * parse the input document.
+ */
+class WP_HTML_Processor {
+	/**
+	 * Describes the constraints imposed on the HTML tag we're trying to match.
+	 *
+	 * @see WP_Tag_Find_Descriptor::parse()
+	 *
+	 * @var WP_Tag_Find_Descriptor
+	 */
+	public $descriptor;
+
+
+	/**
+	 * Scans the HTML document, keeping track of its own state.
+	 *
+	 * @var WP_HTML_Scanner
+	 */
+	public $scanner;
+
+
+	/**
+	 * Set of modifications to perform on input HTML document, built
+	 * via calls to the fluid API exposed by WP_HTML_Processor.
+	 *
+	 * @var WP_Tag_Modifications
+	 */
+	public $modifications;
+
+
+	public function __construct( $input ) {
+		$this->scanner = new WP_HTML_Scanner( $input );
+		$this->modifications = new WP_Tag_Modifications();
+	}
+
+
+	public function find( $query ) {
+		$this->commit_class_changes();
+		$this->descriptor = WP_Tag_Find_Descriptor::parse( $query );
+		if ( false === $this->scanner->scan( $this->descriptor ) ) {
+			return false;
+		}
+
+		return $this;
+	}
+
+
+	public function add_attribute( $name, $value ) {
+		$tag_name = $this->scanner->tag_name;
+		$insert_at = $tag_name->start + $tag_name->length;
+
+		$this->modifications->replacements[] = array(
+			$insert_at,
+			$insert_at,
+			' ' . self::serialize_attribute( $name, $value )
+		);
+
+		return $this;
+	}
+
+
+	public function add_class( $class_name ) {
+		$this->modifications->class_changes[ $class_name ] = true;
```

Comment:
I used a similar data structure in https://github.com/WordPress/gutenberg/pull/42485! :-)